nav-app/src/app/techniques-search/techniques-search.component.ts

```diff
@@ -37,71 +39,140 @@ export class TechniquesSearchComponent implements OnInit {
+                        if (technique[field.field]?.toLowerCase().includes(query.trim().toLowerCase())) {
+                            seenIDs.add(technique.id);
+                            return true;
+                        }
                     }
                 }
                 return false;
             });
-            // deconflict IDs for cross-tactic techniques
-            let seenIDs = new Set();
-            results = results.filter(function(technique: Technique) {
-                if (seenIDs.has(technique.id)) return false;
-                else {
-                    seenIDs.add(technique.id);
-                    return true;
-                }
-            })
+        }
+        return results;
+    }
 
-            results = results.sort((tA: Technique, tB: Technique) => {
-                let c1 = tA.isSubtechnique ? tA.parent.name : tA.name;
-                let c2 = tB.isSubtechnique ? tB.parent.name : tB.name;
-                return c1.localeCompare(c2)
-            });
-            // console.log(seenIDs, results)
-            this.results = results;
-        } else {
-            this.results = [];
+    getTechniques() {
+        //get master list of techniques and sub-techniques
+        let allTechniques = this.dataService.getDomain(this.viewModel.domainID).techniques;
+        for (let technique of allTechniques) {
+            allTechniques = allTechniques.concat(technique.subtechniques);
         }
+        this.techniqueResults = this.filterAndSort(allTechniques, "", true);
     }
-    public get queryLength(): number {
-        return this._query.length;
+
+    getTechniqueResults(query = "") {
+        if (query.trim() != "" && query.includes(this.previousQuery)) {
+            this.techniqueResults = this.filterAndSort(this.techniqueResults, query, true);
+        } else {
+            this.getTechniques();
+        }
     }
-    public results: Technique[] = [];
 
-    constructor(private dataService: DataService) {
+    getStixData() {
+        let domain = this.dataService.getDomain(this.viewModel.domainID);
 
+        this.stixTypes = [{
+            "label": "threat groups",
+            "objects": this.filterAndSort(domain.groups)
+        }, {
+            "label": "software",
+            "objects": this.filterAndSort(domain.software)
+        }, {
+            "label": "mitigations",
+            "objects": this.filterAndSort(domain.mitigations)
+        }]
     }
 
-    ngOnInit() {
+    // getStixResults() checks if query is:
+    // 1) valid, and
+    // 2) part of last query, otherwise call getStixData() to search all objects again
+    getStixResults(query = "") {
+        if (query.trim() != "" && query.includes(this.previousQuery)) {
+            this.stixTypes.forEach(item => item['objects'] = this.filterAndSort(item['objects'], query));
+        } else {
+            this.getStixData();
+        }
```

Comment:
I've not seen optimization like this before. Definitely something I will consider reusing myself in the future!